### PR TITLE
improve proximity

### PR DESCRIPTION
**Pull request by ansis**
This is the carmen-cache side of the proximity improvements.

- make `tileDist` calculation more accurate
    - use euclidean instead of manhattan distance
    - use an unrounded version of the proximity point
- uses a (first attempt at a) better scoredist function

---
 
- [ ] fix tests
- [ ] wait for rest of proximity work in carmen
- [ ] ask for review

### CODE REVIEW

**Comment by mattficke** [comment]
@springmeyer I'm picking this back up, can I get your 👀 /thoughts on this? We're trying to improve proximity biasing for larger features.

At a high level, we're trying to do 2 things:

* calculate distance based on distance to polygon rather than distance to centerpoint (or distance to tile cover when we don't have the feature geometry)
* Once ^^ is working, change the radius of effect so that proximity biasing works from greater distances. This radius will probably vary based on feature type or size.

For context, carmen pr: https://github.com/mapbox/carmen/pull/554

**Comment by springmeyer** [comment]
@mattficke thanks for picking this up and providing some background. My thoughts at this point are only that 1) this looks like an important feature and 2) I can help if you get stuck with any of the C++ details.

**Comment by mattficke** [comment]
Breaking this up into smaller chunks to better test performance implications. Against master, the proximity param has no effect on latency (at least in a gut-check test on staging with geocoder-test). Next up:

* test revised tiledist calculation to make sure it doesn't slow things down

**Comment by mattficke** [comment]
Need to fix tests and do benchmarking, but I think I have the last piece here now. Turns out we were sorting covers by `scoreDist`, but not distance. This effectively meant that we weren't sorting covers by distance unless the proximity point was on the same tile.

I added 59cc785 so that we return the closest tile to the proximity point in cases when there's a tie in `scoredist`. The distance value is already being generated, so there shouldn't be a performance implication (though this needs to be confirmed).

The idea here is that, over in Carmen, we can use distance-to-cover as a cheap approximation of distance-to-poly,. This will replace distance-to-center as the distance measure we use for proximity weighting.

**Comment by mattficke** [comment]
Benchmarking plan:

* compare `coalesce.bench.js` proximity results on this branch vs master
* increase number of runs to better see any variation: current runs are < 8ms
* benchmark `coalesce` function over in carmen with benchmark.js library

**Comment by mattficke** [comment]
The shift to doubles + euclidian distance does seem to slow things down a very small amount, in the range of .2-.4ms (<5%) as measured by the coalesce bench test. Going to keep things as-is for now, will revisit if this is a meaningful amount or if the benchmark in carmen reveals a larger impact.

**Comment by mattficke** [comment]
Tried to revert all the new `doubles` back to `unsigned`, but ran into an error downstream when running a proximity test. Tests here in carmen-cache didn't catch it. @springmeyer, is this something we can/should test for here?

```
buffer.js:615
    throw new RangeError('index out of range');
    ^

RangeError: index out of range
    at checkOffset (buffer.js:615:11)
    at Buffer.readUInt32BE (buffer.js:689:5)
    at /mnt/data/api-geocoder/node_modules/memjs/lib/memjs/memjs.js:278:34
    at responseHandler (/mnt/data/api-geocoder/node_modules/memjs/lib/memjs/memjs.js:491:17)
    at Server.respond (/mnt/data/api-geocoder/node_modules/memjs/lib/memjs/server.js:33:3)
    at Server.responseHandler (/mnt/data/api-geocoder/node_modules/memjs/lib/memjs/server.js:86:12)
    at Socket.<anonymous> (/mnt/data/api-geocoder/node_modules/memjs/lib/memjs/server.js:109:14)
    at emitOne (events.js:77:13)
    at Socket.emit (events.js:169:7)
    at readableAddChunk (_stream_readable.js:153:18)
```

File: src/binding.cpp [hunk 1]
```diff
@@ -1135,18 +1139,19 @@ void coalesceSingle(uv_work_t* req) {
...
             cover.idx = subq.idx;
             cover.tmpid = static_cast<uint32_t>(cover.idx * POW2_25 + cover.id);
             cover.relev = cover.relev * subq.weight;
-            cover.distance = proximity ? tileDist(cx, cover.x, cy, cover.y) : 0;
+            cover.distance = proximity ? tileDist(cx, cy, cover.x, cover.y) : 0;
             cover.scoredist = proximity ? scoredist(cz, cover.distance, cover.score) : cover.score;
 
             // only add cover id if it's got a higer scoredist
-            if (lastId == cover.id && cover.scoredist <= lastScoredist) continue;
+            if (lastId == cover.id && cover.scoredist <= lastScoredist && cover.distance >= lastDistance) continue;
```
Comment (yhahn) [inline src/binding.cpp]: Hrm, what's the intention here? Seems like the purpose of scoredist is to abstract away the score/dist individual checks, so why would we need an additional distance check here?

Comment (mattficke) [inline src/binding.cpp]: @yhahn this is the change that allows the closest tile to the proximity point to be returned. In many cases, if a tile is more than a couple of squares away on the z/x/y grid, `scoredist` will be the same even if the actual distance is quite different.

File: src/binding.cpp [hunk 2]
```diff
@@ -1042,12 +1046,14 @@ struct CoalesceBaton : carmen::noncopyable {
...
-    if (zoom == 10) scoredist = 2.0 / distance;
-    if (zoom <= 9)  scoredist = 1.0 / distance;
+    if (zoom >= 13) scoredist = 32.0 / distance;
+    if (zoom == 12) scoredist = 24.0 / distance;
+    if (zoom == 11) scoredist = 16.0 / distance;
+    if (zoom == 10) scoredist = 10.0 / distance;
+    if (zoom == 9)  scoredist = 6.0 / distance;
+    if (zoom == 8)  scoredist = 3.5 / distance;
+    if (zoom == 7)  scoredist = 2.0 / distance;
+    if (zoom <= 6)  scoredist = 1.125 / distance;
```
Comment (mattficke) [inline src/binding.cpp]: The scoredist calc here now matches the carmen changes, and scales the radius of effect based on zoom level.

**Pull request merged**